lib/location/method_cellular.c

```diff
@@ -136,11 +141,33 @@ static void method_cellular_positioning_work_fn(struct k_work *work)
 	/* NCELLMEAS done at this point of time. Store current time to response. */
 	location_utils_systime_to_location_datetime(&location_result.datetime);
 
+	/* Check if timeout is given */
+	if (cellular_config.timeout > 0) {
+		/* +1 to round the time up */
+		ncellmeas_time = ((k_uptime_get() - ncellmeas_start_time) / MSEC_PER_SEC) + 1;
+
+		/* Check if timeout has already elapsed */
+		if (ncellmeas_time >= cellular_config.timeout) {
+			LOG_WRN("Timeout occurred during neighbour cell measurement");
+			location_core_event_cb_timeout();
+			running = false;
+			return;
+		}
+		/* Take time used for neighbour cell measurements into account */
+		params.timeout = cellular_config.timeout - ncellmeas_time;
```

Comment:
To be more specific on this, when location_core_timeout timer triggers, it can send location event and indicate that timeout occurred but `multicell_location_get()` will remain running in the background. This will take us into a state where new location_request() cannot be sent until `multicell_location_get()` has really finalized. And client cannot really know when this would be.

Reply:
Then, you are right, my suggestion would not work. I did not realize that the `http_client` implementation blocks during the request. I assumed the calling thread was yielding while waiting for response.

I ***think*** `http_client` could be modified to be non-blocking. For instance, checking for `EAGAIN` inside `http_wait_data`, and yielding the calling thread for a short time if that occurs, rather than simply failing. Then we could use a non-blocking socket in the REST client.

If we did that, then it would be possible to rely solely on the `location_core` timeout.

Not sure whether pursuing that approach is worth waiting for the required modification to `http_client`.  Additionally, this is getting into fairly low-level socket behavior, though, and I am not an expert, so I may just be outright wrong about whether it is even possible for `http_client` to yield the calling thread rather than block it.

If it really is not possible to use the `location_core` timeout, then I wonder whether we should have it at all. As I said in my other comment, I think it would be cleaner to have only a single timeout mechanism running at any given point in time. If we are already being forced to implement timeout logic individually in each of the service integrations, we might as well rely on it solely, no?

Reply:
First some comments generally on what the problems are at the moment. You wrote in comment https://github.com/nrfconnect/sdk-nrf/pull/7978#discussion_r905395148:
> In fact, I am pretty sure that if cell measurement finishes at just the right time, it will be a coin-toss between whether this work function resumes on line 129, or the the location_core timeout work function initiates. That potential for inconsistency makes me nervous.

This is true. However, right now without this PR, we have much worse problems. HTTP request can be pending for a long time (seconds and tens of seconds). Location core timeout expires and the API returns but new requests cannot really be made as the lower layers are blocked.
This PR is not fixing everything and there are holes when this might still occur. However, time window for these issues to occur will reduce quite a lot through this PR, which is why I'd like to get this done (this week before our holiday season) even if this would be an intermediate step. If we get updates to `http_client` in the future, we are more than happy to update `rest_client`, `multicell_location` and `location` libraries to take advantage of those changes.

I don't disagree w.r.t two timeouts being complicated. Lower layers are not really our friends at the moment to implement nice timeout mechanism with one timeout so we are compromising somewhere anyway.
Using non-blocking sockets could work but I'm not 100% sure about that either.
In addition, I'd like to have HTTP timeout for the entire operation including the sending part, not just receiving.

I don't know yet if we could just disable location core timeout after ncellmeas has finished and just rely on `http_client` and socket timeouts.

Reply:
> I don't know yet if we could just disable location core timeout after ncellmeas has finished and just rely on `http_client` and socket timeouts.

I've done changes to stop core timer after ncellmeas and wifi scanning.

Reply:
Thanks for hearing out and addressing my concerns. I agree that we are kind of between a rock and a hard place due to how `http_client` works. I'm frankly not 100% sure whether we should choose the rock, or the hard place, but that's ultimately your decision as a code owner, not mine. I'll go ahead and remove my request for changes.